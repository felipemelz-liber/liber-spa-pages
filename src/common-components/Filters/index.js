export { default as ArrayFieldFilter } from './ArrayFieldFilter/ArrayFieldFilter';
export { default as TextFieldFilter } from './TextFieldFilter/TextFieldFilter';
export { default as DateFieldFilter } from './DateFieldFilter/DateFieldFilter';
