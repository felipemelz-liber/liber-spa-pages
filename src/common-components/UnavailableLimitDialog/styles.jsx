import styled from 'styled-components';
import { FlatButton } from 'liber-components';

const TEXT_COLOR = '#6287a7';

export const Container = styled.div`
  padding: 24px 24px 12px 24px;
  width: 480px;

  * {
    font-family: Roboto;
  }
`;

export const Header = styled.div`
  font-size: 21px;
  font-weight: bold;
  line-height: 1.4;
  letter-spacing: 0.08px;
  color: ${TEXT_COLOR};
  margin-bottom: 24px;
`;

export const Content = styled.div`
  font-size: 16px;
  line-height: 1.38;
  letter-spacing: 0.01px;
  margin-bottom: 24px;
  color: ${TEXT_COLOR};
  margin-bottom: 24px;
`;

export const ActionsRow = styled.div`
  display: flex;
  justify-content: flex-end;
`;

export const Button = styled(FlatButton).attrs(() => ({ size: 'large' }))``;
