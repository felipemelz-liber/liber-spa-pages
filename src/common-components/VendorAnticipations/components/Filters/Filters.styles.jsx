import styled from 'styled-components';
import { Button, handleThemeFromObject as getTheme } from 'liber-components';

export const FilterText = styled.span`
  font-family: Roboto;
  font-size: 16px;
  font-weight: normal;
  font-stretch: normal;
  font-style: normal;
  line-height: 1.38;
  letter-spacing: 0.06px;
  text-align: left;
  color: #405f71;
  margin-right: 16px;
`;

export const CleanFilters = styled(Button).attrs(() => ({ version: '2' }))`
  font-weight: 400;
  font-size: 16px;
  line-height: 22px;
  letter-spacing: 0.1px;
  color: ${({ theme }) => getTheme(theme, 'colors.primary', '#009dff')};
`;
