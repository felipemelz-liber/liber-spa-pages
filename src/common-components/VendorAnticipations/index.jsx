import React from 'react';
import PropTypes from 'prop-types';
import { Provider } from 'react-redux';
import hydrateStore from '../../views/store';
import VendorAdvances from './components/VendorAdvances';

const VendorAnticipations = ({ token, vendorId, operator, isAdmin, onDetailsClick }) => (
  <>
    <Provider store={hydrateStore({ vendorAdvances: { token, vendorId, operator, isAdmin } })}>
      <VendorAdvances onDetailsClick={onDetailsClick} />
    </Provider>
  </>
);

VendorAnticipations.propTypes = {
  token: PropTypes.string.isRequired,
  vendorId: PropTypes.number.isRequired,
  operator: PropTypes.shape({
    id: PropTypes.number,
    name: PropTypes.string,
    cpf: PropTypes.string,
  }).isRequired,
  isAdmin: PropTypes.bool,
  onDetailsClick: PropTypes.func,
};

VendorAnticipations.defaultProps = {
  isAdmin: false,
  onDetailsClick: () => null,
};

export default VendorAnticipations;
