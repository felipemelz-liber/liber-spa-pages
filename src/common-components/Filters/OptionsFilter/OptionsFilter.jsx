import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import FilterTag from '../FilterTag';
import SelectTag from '../../SelectTag/SelectTag';
import FilterFooter from '../FilterFooter';
import { OptionsContainer } from './OptionsFilter.styles';

const handleFilter = (option, selectedState) => () => {
  const [selectedOptions, setSelectedOptions] = selectedState;
  let finalSelection = [];

  if (selectedOptions.includes(option)) {
    finalSelection = selectedOptions.filter(selectedOption => selectedOption !== option);
  } else {
    finalSelection = [...selectedOptions, option];
  }

  setSelectedOptions(finalSelection);
};

const formatLabel = label => list => {
  const [selectedOptions] = list;

  return `${label} (${selectedOptions.length})`;
};

const OptionsFilter = ({ options, onFilter, onClear, filterValues, label }) => {
  const [open, setOpen] = useState(false);
  const selectedState = useState(filterValues || []);
  const [selectedOptions, setSelectedOptions] = selectedState;

  useEffect(() => {
    if (!filterValues) {
      setSelectedOptions([]);
    }
  }, [filterValues]);

  return (
    <FilterTag
      selected={[filterValues]}
      onClear={onClear}
      show={open}
      label={label}
      formatSelected={formatLabel(label)}
      onChange={setOpen}
    >
      <OptionsContainer>
        {options.map(({ value, option }) => (
          <SelectTag
            key={`option-${value}`}
            onChange={handleFilter(value, selectedState)}
            selected={selectedOptions.includes(value)}
          >
            {option}
          </SelectTag>
        ))}
      </OptionsContainer>
      <FilterFooter
        disabled={selectedOptions.length === 0}
        onFilter={() => {
          setOpen(false);
          onFilter(selectedOptions);
        }}
        onClear={onClear}
      />
    </FilterTag>
  );
};

OptionsFilter.propTypes = {
  options: PropTypes.arrayOf(PropTypes.shape({ value: PropTypes.any, option: PropTypes.string })),
  filterValues: PropTypes.arrayOf(PropTypes.any),
  onFilter: PropTypes.func,
  onClear: PropTypes.func,
  label: PropTypes.string,
};

OptionsFilter.defaultProps = {
  options: [],
  filterValues: null,
  onFilter: () => {},
  onClear: () => {},
  label: '',
};

export default OptionsFilter;
