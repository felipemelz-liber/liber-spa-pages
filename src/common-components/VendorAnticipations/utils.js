/* eslint-disable camelcase */
import React from 'react';

import { toast } from 'liber-components/components/Toast';
import { camelCase, find } from 'lodash';
import { formatMoney } from '../../views/withdraw/moneyHandler';
import StatusText from '../StatusText/StatusText';
import { EyeIcon } from './components/PendingAdvances/PendingAdvances.styles';
import { dispatchDataLayerEvent } from '../../vendor/Utils';

export const ADVANCE_REQUESTED = 'Assinatura Pendente';
export const ADVANCE_WAITING_PAYMENT = 'Em Processamento';
export const ADVANCE_CONCLUDED = 'Concluídos';
export const ADVANCE_CANCELED = 'Cancelados';
export const ADVANCE_ALL = 'Todos';

export const TOAST_DURATION = 5000;

export const DUPLICATA_NUMBER_IN = 'q[duplicata_number_cont_any]';
export const STATUS_EQ = 'q[status_eq]';
export const STATUS_IN = 'q[status_in]';
export const REQUESTED_GT = 'q[created_at_gteq]';
export const REQUESTED_LT = 'q[created_at_lteq]';

export const STATUS = {
  created: 0,
  started: 1,
  signed: 2,
  finished: 3,
  canceled: 4,
};

export const STATUS_FILTERS_TYPES = {
  concluded: {
    [STATUS_IN]: [STATUS.finished, STATUS.canceled],
  },
  pending: {
    [STATUS_IN]: [STATUS.started, STATUS.signed, STATUS.created],
  },
};

export const TABS = [
  ADVANCE_REQUESTED,
  ADVANCE_WAITING_PAYMENT,
  ADVANCE_CONCLUDED,
  ADVANCE_CANCELED,
  ADVANCE_ALL,
];

const emptyCell = '***';

export const STATUS_TEXTS = {
  started: 'Aguardando assinatura',
  signed: 'Realizando Pagamento',
  finished: 'Concluída',
  canceled: 'Cancelada',
};

export const convertToCamelCase = object => {
  const convertedObject = {};
  Object.keys(object).forEach(key => {
    const newKey = camelCase(key);
    convertedObject[newKey] = object[key];
  });

  return convertedObject;
};

export const getStatusCount = (documents, status) =>
  documents.reduce((totalStatus, document) => {
    const { status: documentStatus } = document;
    if (documentStatus === status) {
      return totalStatus + 1;
    }
    return totalStatus;
  }, 0);

export const updateAdvances = (advances, documents) =>
  advances.map(advance => {
    const { id } = advance;
    const updateObject = find(documents, { id });

    if (updateObject) {
      const { signingId } = updateObject;
      return { ...advance, signingId };
    }

    return advance;
  });

export const updateSignedAdvances = (advances, documents) =>
  advances.map(advance => {
    const { id } = advance;
    const updateObject = find(documents, { id });

    if (updateObject) {
      const { status, signatureProgress } = updateObject;

      return { ...advance, signatureProgress, userHasSigned: status !== 'error' };
    }

    return advance;
  });

export const calculateQuote = advances => {
  const finalQuote = advances.reduce(
    (quote, current) => {
      const faceValue = quote.faceValue + parseFloat(current.faceValue);
      const discountRate =
        quote.discountRate + parseFloat(current.agreedRate) * current.invoiceCount;
      const invoiceCount = quote.invoiceCount + parseFloat(current.invoiceCount);
      const netValue = quote.netValue + parseFloat(current.netValue);

      return { faceValue, invoiceCount, netValue, discountRate };
    },
    { faceValue: 0, discountRate: 0, invoiceCount: 0, netValue: 0 },
  );

  const { faceValue, invoiceCount, netValue, discountRate: total } = finalQuote;
  const discountRate = total / invoiceCount;

  return {
    faceValue,
    discountRate,
    netValue,
    invoiceCount,
  };
};

export const handleError = () => {
  toast(
    {
      message: 'Um erro ocorreu',
      info: 'Por favor tente novamente mais tarde ou entre em contato com o suporte',
    },
    'error',
    TOAST_DURATION,
  );
};

export const handleSignatureFeedback = (errorCount, completedCount, documentsCount) => {
  if (errorCount > 0) {
    if (completedCount > 0) {
      toast(
        {
          message: 'Ops.. Algumas assinaturas não foram realizadas!',
          info: 'Veja as assinaturas pendentes e tente novamente',
        },
        'warning',
        TOAST_DURATION,
      );
    } else if (documentsCount - errorCount !== completedCount) {
      toast(
        {
          message: 'Ops.. Algumas assinaturas não foram realizadas!',
          info: 'Tente novamente ou fale com o suporte',
        },
        'warning',
        TOAST_DURATION,
      );
    } else {
      toast(
        {
          message: 'Erro, sua assinatura não foi realizada!',
          info: 'Tente novamente ou fale com o suporte',
        },
        'error',
        TOAST_DURATION,
      );
    }
  } else if (completedCount > 0) {
    toast(
      {
        message: 'Termo assinado com sucesso!',
        info: 'Aguarde o processamento do pagamento',
      },
      'success',
      TOAST_DURATION,
    );
  } else {
    toast(
      {
        message: 'Termo assinado com sucesso!',
        info: 'Aguarde as outras assinaturas pendentes',
      },
      'success',
      TOAST_DURATION,
    );
  }
};

const handleAdvanceShowClick = isAdmin => () => {
  dispatchDataLayerEvent('antecipacoesHistoricoDeAntecipacoesShowDoBordero', isAdmin);
};

export const getColumns = (isAdmin, onDetailsClick) => [
  {
    name: 'status',
    label: 'Status',
    width: 14,
    align: 'left',

    getCellValue: advance => {
      const { status } = advance;
      const statusMap = {
        finished: 'positive',
        canceled: 'negative',
      };

      return <StatusText type={statusMap[status]}>{STATUS_TEXTS[status]}</StatusText>;
    },
  },
  {
    name: 'netValue',
    label: 'Valor Antecipado (R$)',
    width: 18,
    align: 'right',
    getCellValue: ({ netValue, status }) =>
      netValue && status !== 'canceled' ? formatMoney(netValue) : emptyCell,
  },
  {
    name: 'invoiceCount',
    label: 'Títulos Solicitados',
    width: 15,
    align: 'right',
  },
  {
    name: 'invoicesAnticipated',
    label: 'Títulos Antecipados',
    width: 15,
    align: 'right',
    getCellValue: ({ paidInvoicesCount }) => paidInvoicesCount || emptyCell,
  },
  {
    name: 'requestedAt',
    label: 'Solicitado Em',
    width: 18,
    align: 'center',
  },
  {
    name: 'documentUrl',
    label: 'Ações',
    width: 8,
    align: 'center',
    // eslint-disable-next-line
    getCellValue: ({ id }) =>
      isAdmin ? (
        <EyeIcon onClick={() => onDetailsClick(id)} />
      ) : (
        <a href={`/fornecedor/antecipacoes/${id}`} onClick={handleAdvanceShowClick(isAdmin)}>
          <EyeIcon />
        </a>
      ),
  },
];
