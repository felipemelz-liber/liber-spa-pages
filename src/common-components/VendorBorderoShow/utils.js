import React from 'react';
import { mdiEye } from '@mdi/js';
import StatusText from '../StatusText/StatusText';
import { formatMoney } from '../../views/withdraw/moneyHandler';
import { StyledIcon } from './components/VendorBorderoShow.styles';
import { dispatchDataLayerEvent } from '../../vendor/Utils';

export const DUPLICATA_NUMBER_IN_BORDERO = 'q[number_in]';
export const VENDOR_CNPJ_EQ = 'q[emitter_cnpj_eq]';
export const BORDERO_STATUS_IN = 'q[status_in]';

const emptyCell = '***';

export const BORDERO_STATUS_MAP = {
  created: {
    label: 'Aguardando Confirmação',
    type: 'neutral',
  },
  started: {
    label: 'Aguardando Confirmação',
    type: 'neutral',
  },
  signed: {
    label: 'Realizando Pagamento',
    type: 'neutral',
  },
  finished: {
    label: 'Concluída',
    type: 'positive',
  },
  canceled: {
    label: 'Cancelada',
    type: 'negative',
  },
};

export const STATUS_REFERENCE = {
  requested: 'Solicitado',
  anticipated: 'Antecipado',
  notAnticipated: 'Não Antecipado',
};

export const INVOICE_STATUS_MAP = {
  aberta: {
    code: 0,
    label: 'Solicitado',
    type: 'neutral',
  },
  aguardando_aceite: {
    code: 1,
    label: 'Solicitado',
    type: 'neutral',
  },
  aguardando_pagamento: {
    code: 3,
    label: 'Solicitado',
    type: 'neutral',
  },
  paga: {
    code: 4,
    label: 'Solicitado',
    type: 'neutral',
  },
  concluida: {
    code: 5,
    label: 'Antecipado',
    type: 'positive',
  },
  cancelada: {
    code: 6,
    label: 'Não Antecipado',
    type: 'negative',
  },
  agendada: {
    code: 7,
    label: 'Solicitado',
    type: 'neutral',
  },
  aguardando_formalizacao: {
    code: 8,
    label: 'Solicitado',
    type: 'neutral',
  },
  em_processamento: {
    code: 9,
    label: 'Solicitado',
    type: 'neutral',
  },
};

const handleShowInvoiceClick = (setInvoiceId, setInvoiceShowOpen, id, isAdmin) => {
  dispatchDataLayerEvent('showDoBorderoShowDoTitulo', isAdmin);

  setInvoiceId(id);
  setInvoiceShowOpen(true);
};

export const getInvoicesColumns = ({
  borderoStatus,
  setInvoiceId,
  setInvoiceShowOpen,
  isAdmin,
}) => {
  return [
    {
      name: 'status',
      label: <>&nbsp;&nbsp;&nbsp;Status</>,
      width: 17,
      align: 'left',
      getCellValue: invoice => {
        const { type, label } = INVOICE_STATUS_MAP[invoice.status] || INVOICE_STATUS_MAP.aberta;
        return (
          <>
            &nbsp;&nbsp;&nbsp;
            <StatusText type={type}>{label}</StatusText>
          </>
        );
      },
    },
    {
      name: 'number',
      label: 'Número',
      width: 15,
      align: 'right',
    },
    {
      name: 'netValue',
      label: `Valor ${
        ['finished', 'canceled'].includes(borderoStatus) ? 'Antecipado (R$)' : 'Solicitado (R$)'
      }`,
      width: 16,
      align: 'right',
      getCellValue: ({ netValue, status }) =>
        netValue && status !== 'cancelada' ? formatMoney(netValue) : emptyCell,
    },
    {
      name: 'vendorCnpj',
      label: 'CNPJ',
      width: 15,
      align: 'right',
    },
    {
      name: 'funder',
      label: 'Investidor',
      width: 14,
      align: 'left',
      getCellValue: ({ funder, status }) => (funder && status !== 'cancelada' ? funder : emptyCell),
    },
    {
      name: 'detailsPath',
      label: 'Ações',
      width: 7,
      align: 'center',
      // eslint-disable-next-line
      getCellValue: ({ id }) => (
        <StyledIcon
          title="Visualizar título"
          path={mdiEye}
          onClick={() => handleShowInvoiceClick(setInvoiceId, setInvoiceShowOpen, id, isAdmin)}
        />
      ),
    },
  ];
};

export const COLUMNS_BORDERO_INVOICES = [
  {
    name: 'status',
    label: <>&nbsp;&nbsp;&nbsp;Status</>,
    width: 17,
    align: 'left',
    getCellValue: invoice => {
      const { type, label } = INVOICE_STATUS_MAP[invoice.status] || INVOICE_STATUS_MAP.aberta;
      return (
        <>
          &nbsp;&nbsp;&nbsp;
          <StatusText type={type}>{label}</StatusText>
        </>
      );
    },
  },
  {
    name: 'number',
    label: 'Número',
    width: 15,
    align: 'right',
  },
  {
    name: 'netValue',
    label: 'Valor Antecipado (R$)',
    width: 16,
    align: 'right',
    getCellValue: ({ netValue, status }) =>
      netValue && status !== 'cancelada' ? formatMoney(netValue) : emptyCell,
  },
  {
    name: 'vendorCnpj',
    label: 'CNPJ',
    width: 15,
    align: 'right',
  },
  {
    name: 'funder',
    label: 'Investidor',
    width: 14,
    align: 'left',
    getCellValue: ({ funder }) => funder || emptyCell,
  },
  {
    name: 'detailsPath',
    label: 'Ações',
    width: 7,
    align: 'center',
    // eslint-disable-next-line
    getCellValue: ({ detailsPath = '#' }) => (
      <StyledIcon
        title="Visualizar título"
        path={mdiEye}
        onClick={() => {
          window.location.href = detailsPath;
        }}
      />
    ),
  },
];

export const getCodesByLabels = labels => {
  const statusCodes = [];
  Object.keys(INVOICE_STATUS_MAP).forEach(key => {
    const { label, code } = INVOICE_STATUS_MAP[key];
    if (labels.includes(label)) {
      statusCodes.push(code);
    }
  });
  return statusCodes;
};

export const formatFilters = filters => {
  const newFilters = {};
  Object.keys(filters).forEach(key => {
    if (key === BORDERO_STATUS_IN) {
      newFilters[BORDERO_STATUS_IN] = getCodesByLabels(
        filters[BORDERO_STATUS_IN]
          ? filters[BORDERO_STATUS_IN].map(item => STATUS_REFERENCE[item])
          : [],
      );
    } else {
      newFilters[key] = filters[key];
    }
  });
  return newFilters;
};

export const MOCKED_SUMMARY = {
  // status: 'finished',
  // status: 'signed',
  status: 'started',
  buyer: { name: 'BRF S.A.', cnpj: '06.194.174/0001-19' },
  paid_value: 'R$ 1.018.106,59',
  invoice_count: '13347',
  paid_invoice_count: undefined,
  requested_at: '10/02/2020',
  operators: [
    {
      id: 2,
      name: 'Jamile Cardoso Gomes de Jesus',
      user_has_signed: true,
    },
    {
      id: 1,
      name: 'Zezinho',
      user_has_signed: false,
    },
  ],
  bank_account: {
    account: '2056',
    account_cd: '2',
    branch: '0004',
    branch_cd: '1',
    bank_name: 'Banco Bradesco S.A.',
    id: 5,
    bank_code: '237',
  },
  signature_certificate_url: '#1', // null => nao aparece o botao
  invoice_transfer_agreement_url: '#2',
};

export const MOCKED_INVOICES = [
  {
    id: 1,
    status: 'em_processamento',
    number: '000003665-20-001',
    net_value: 3456.42,
    vendor_cnpj: '06.194.174/0001-19',
    funder: 'BTG Pactual',
    details_path: '/fornecedor/duplicatas/dc1541250297',
  },
  {
    id: 2,
    status: 'cancelada',
    number: '000003665-20-001',
    net_value: 296.53,
    vendor_cnpj: '06.194.174/0001-19',
    funder: 'BTG Pactual',
    details_path: '/fornecedor/duplicatas/dc1541250297',
  },
  {
    id: 3,
    status: 'concluida',
    number: '000003665-20-001',
    net_value: 2416.39,
    vendor_cnpj: '06.194.174/0001-19',
    funder: 'BTG Pactual',
    details_path: '/fornecedor/duplicatas/dc1541250297',
  },
  {
    id: 4,
    status: 'em_processamento',
    number: '000003665-20-001',
    net_value: 3456.42,
    vendor_cnpj: '06.194.174/0001-19',
    funder: 'BTG Pactual',
    details_path: '/fornecedor/duplicatas/dc1541250297',
  },
  {
    id: 5,
    status: 'cancelada',
    number: '000003665-20-001',
    net_value: 296.53,
    vendor_cnpj: '06.194.174/0001-19',
    funder: 'BTG Pactual',
    details_path: '/fornecedor/duplicatas/dc1541250297',
  },
  {
    id: 6,
    status: 'concluida',
    number: '000003665-20-001',
    net_value: 2416.39,
    vendor_cnpj: '06.194.174/0001-19',
    funder: 'BTG Pactual',
    details_path: '/fornecedor/duplicatas/dc1541250297',
  },
];
