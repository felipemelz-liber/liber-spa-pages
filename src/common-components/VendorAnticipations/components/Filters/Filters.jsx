import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import DateFieldFilter from '../../../Filters/DateFieldFilter/DateFieldFilter';
import { REQUESTED_LT, REQUESTED_GT, STATUS, STATUS_IN } from '../../utils';
import { updateFilter } from '../../actions/VendorAdvancesActions';
import OptionsFilter from '../../../Filters/OptionsFilter/OptionsFilter';
import { FilterText, CleanFilters } from './Filters.styles';
import { dispatchDataLayerEvent } from '../../../../vendor/Utils';

const clearDateFilters = {
  [REQUESTED_LT]: null,
  [REQUESTED_GT]: null,
};

const clearStatusFilters = {
  [STATUS_IN]: null,
};

const allClearFilters = {
  ...clearDateFilters,
  ...clearStatusFilters,
};

const onDateFilter = (changeFilters, applyedFilters, isAdmin) => range => {
  dispatchDataLayerEvent('antecipacoesFiltrarSolicitacao', isAdmin);

  let filters = {};

  if (range && range.end) {
    filters = {
      [REQUESTED_GT]: range.start.format(),
      [REQUESTED_LT]: range.end.format(),
    };
  } else {
    filters = {
      [REQUESTED_GT]: range && range.start ? range.start.startOf('day').format() : null,
      [REQUESTED_LT]: range && range.start ? range.start.endOf('day').format() : null,
    };
  }

  changeFilters({ ...applyedFilters, ...filters });
};

const onStatusFilter = (filters, changeFilters, selectedStatuses, isAdmin) => {
  const status = selectedStatuses
    .map(statusId => (statusId === STATUS.finished ? 'Concluida' : 'Cancelada'))
    .join(' | ');

  dispatchDataLayerEvent('antecipacoesFiltrarStatus', isAdmin, {
    filtroStatusAntecipacoes: status,
  });

  changeFilters({ ...filters, [STATUS_IN]: selectedStatuses });
};

const handleDateFilterClear = (changeFilters, filters, isAdmin) => {
  dispatchDataLayerEvent('antecipacoesLimparFiltro', isAdmin);

  changeFilters({ ...filters, ...clearDateFilters });
};

const handleStatusFilterClear = (changeFilters, filters, isAdmin) => {
  dispatchDataLayerEvent('antecipacoesLimparFiltro', isAdmin);

  changeFilters({ ...filters, ...clearStatusFilters });
};

const handleClearAllFiltersClick = (changeFilters, isAdmin) => {
  dispatchDataLayerEvent('antecipacoesLimparFiltro', isAdmin);

  changeFilters(allClearFilters);
};

const hasFilters = filters =>
  Object.keys(filters).reduce((flag, key) => flag || Boolean(filters[key]), false);

export const Filters = ({ changeFilters, filters, isAdmin }) => {
  return (
    <>
      <FilterText>Filtrar por</FilterText>
      <DateFieldFilter
        label="Solicitação"
        onFilter={onDateFilter(changeFilters, filters, isAdmin)}
        onClear={() => handleDateFilterClear(changeFilters, filters, clearDateFilters, isAdmin)}
        selectedItems={[filters[REQUESTED_GT], filters[REQUESTED_LT]]}
      />
      <OptionsFilter
        filterValues={filters[STATUS_IN]}
        onFilter={selectedStatuses =>
          onStatusFilter(filters, changeFilters, selectedStatuses, isAdmin)
        }
        onClear={() => handleStatusFilterClear(changeFilters, filters, clearStatusFilters, isAdmin)}
        options={[
          { option: 'Concluída', value: STATUS.finished },
          { option: 'Cancelada', value: STATUS.canceled },
        ]}
        label="Status"
      />
      {hasFilters(filters) ? (
        <CleanFilters onClick={() => handleClearAllFiltersClick(changeFilters, isAdmin)}>
          Limpar filtros
        </CleanFilters>
      ) : null}
    </>
  );
};

Filters.propTypes = {
  changeFilters: PropTypes.func,
  filters: PropTypes.shape({
    [REQUESTED_LT]: PropTypes.string,
    [REQUESTED_GT]: PropTypes.string,
    [STATUS_IN]: PropTypes.arrayOf(PropTypes.number),
  }),
  isAdmin: PropTypes.bool,
};

Filters.defaultProps = {
  filters: {},
  isAdmin: false,
  changeFilters: () => {},
};

const mapStateToProps = ({ vendorAdvances: { filters, isAdmin } }) => ({ filters, isAdmin });

const mapDispatchToProps = {
  changeFilters: updateFilter,
};

export default connect(mapStateToProps, mapDispatchToProps)(Filters);
