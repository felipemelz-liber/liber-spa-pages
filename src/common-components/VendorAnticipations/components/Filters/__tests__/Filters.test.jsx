import React from 'react';
import sinon from 'sinon';
import moment from 'moment';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import { Filters } from '../Filters';
import { REQUESTED_LT, REQUESTED_GT, STATUS_IN } from '../../../utils';
import { DateFieldFilter } from '../../../../../common-components/Filters';
import OptionsFilter from '../../../../../common-components/Filters/OptionsFilter/OptionsFilter';
import { CleanFilters } from '../Filters.styles';

const clearDateFilters = {
  [REQUESTED_LT]: null,
  [REQUESTED_GT]: null,
};

const clearStatusFilters = {
  [STATUS_IN]: null,
};

const allClearFilters = {
  ...clearDateFilters,
  ...clearStatusFilters,
};

describe('Filters tests', () => {
  configure({ adapter: new Adapter() });

  const changeFilters = sinon.spy();

  const props = {
    changeFilters,
    filters: allClearFilters,
  };

  const filters = {
    ...allClearFilters,
    [STATUS_IN]: [1, 2],
  };

  const rangeSingle = { start: moment(), end: null };

  const rangePeriod = { start: moment(), end: moment() };

  const periodDateResult = {
    [REQUESTED_LT]: rangePeriod.end.format(),
    [REQUESTED_GT]: rangePeriod.start.format(),
  };
  const singleDateResult = {
    [REQUESTED_GT]: rangeSingle.start.startOf('day').format(),
    [REQUESTED_LT]: rangeSingle.start.endOf('day').format(),
  };

  afterEach(() => {
    sinon.resetHistory();
  });

  it('should render correctly without props', () => {
    const wrapper = shallow(<Filters />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should call change filters on date filter clear', () => {
    const wrapper = shallow(<Filters {...props} />);
    const filter = wrapper.find(DateFieldFilter);

    filter.simulate('clear');

    expect(changeFilters.calledWith({ ...props.filters, ...clearDateFilters })).toBeTruthy();
  });

  it('should call change filters on options filter clear', () => {
    const wrapper = shallow(<Filters {...props} />);
    const filter = wrapper.find(OptionsFilter);

    filter.simulate('clear');

    expect(changeFilters.calledWith({ ...props.filters, ...clearStatusFilters })).toBeTruthy();
  });

  it('should call change filters on clear all button click', () => {
    const wrapper = shallow(<Filters {...props} filters={filters} />);
    const button = wrapper.find(CleanFilters);

    button.simulate('click');

    expect(changeFilters.calledWith(allClearFilters)).toBeTruthy();
  });

  it('should call change filters correctly on option filter on filter', () => {
    const wrapper = shallow(<Filters {...props} />);
    const filter = wrapper.find(OptionsFilter);

    filter.simulate('filter', [1, 2]);

    expect(changeFilters.calledWith({ ...allClearFilters, [STATUS_IN]: [1, 2] })).toBeTruthy();
  });

  it('should call change filters correctly on date filter onFilter with single date', () => {
    const wrapper = shallow(<Filters {...props} />);
    const filter = wrapper.find(DateFieldFilter);

    filter.simulate('filter', rangeSingle);

    expect(changeFilters.calledWith({ ...allClearFilters, ...singleDateResult })).toBeTruthy();
  });

  it('should call change filters correctly on date filter onFilter with period date', () => {
    const wrapper = shallow(<Filters {...props} />);
    const filter = wrapper.find(DateFieldFilter);

    filter.simulate('filter', rangePeriod);

    expect(changeFilters.calledWith({ ...allClearFilters, ...periodDateResult })).toBeTruthy();
  });
});
