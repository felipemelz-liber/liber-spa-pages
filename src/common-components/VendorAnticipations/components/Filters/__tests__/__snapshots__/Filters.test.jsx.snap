// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filters tests should render correctly without props 1`] = `
<Fragment>
  <styled.span>
    Filtrar por
  </styled.span>
  <DateFieldFilter
    description="Selecione uma data de início e uma data de fim para o período desejado:"
    label="Solicitação"
    onClear={[Function]}
    onFilter={[Function]}
    selectedItems={
      Array [
        undefined,
        undefined,
      ]
    }
  />
  <OptionsFilter
    filterValues={null}
    label="Status"
    onClear={[Function]}
    onFilter={[Function]}
    options={
      Array [
        Object {
          "option": "Concluída",
          "value": 3,
        },
        Object {
          "option": "Cancelada",
          "value": 4,
        },
      ]
    }
  />
</Fragment>
`;
